### PR TITLE
Add embeddinator support.

### PR DESCRIPTION
**Pull request by rolfbjarne**
This is a series of commits that adds an embeddinator mode to mtouch and our
runtime.

The commits are quite small and self-contained, but they don't really make
much sense by themselves (separate from embeddinator support), so I put them
all in one pull request.

It's probably easiest to review commit-by-commit, since the commit message explains each change.

### CODE REVIEW

File: tools/linker/MobileResolveMainAssemblyStep.cs [hunk 1]
```diff
@@ -32,8 +35,16 @@ protected override void Process ()
...
+				if (t.IsPublic && embeddinator) {
+					// Mark all public types when in embeddinator mode.
+					// There may be no types with the Register attribute,
+					// which means that without this the assembly might be completely ignored even if it's not linked.
+					Annotations.Mark (t);
 					continue;
+				}
+
+				if (!t.HasCustomAttribute (Namespaces.Foundation, "RegisterAttribute"))
+						continue;
```
Comment (spouliot) [inline tools/linker/MobileResolveMainAssemblyStep.cs]: extra tab ;-)

File: tools/linker/MobileResolveMainAssemblyStep.cs [hunk 2]
```diff
@@ -32,8 +35,16 @@ protected override void Process ()
 			Annotations.Push (assembly);
 
 			foreach (var t in assembly.MainModule.Types) {
-				if (!t.HasCustomAttribute (Namespaces.Foundation, "RegisterAttribute"))
+				if (t.IsPublic && embeddinator) {
+					// Mark all public types when in embeddinator mode.
+					// There may be no types with the Register attribute,
+					// which means that without this the assembly might be completely ignored even if it's not linked.
+					Annotations.Mark (t);
```
Comment (spouliot) [inline tools/linker/MobileResolveMainAssemblyStep.cs]: not sure it's (all) needed - but we can revisit that later (no harm in extra annotations)

File: runtime/xamarin/runtime.h [hunk 3]
```diff
@@ -109,6 +115,7 @@ struct AssemblyLocations {
 };
 
 void xamarin_initialize ();
+void xamarin_initialize_embedded (); /* STABLE */
```
Comment (spouliot) [inline runtime/xamarin/runtime.h]: `/* STABLE */` ?

Comment (rolfbjarne) [inline runtime/xamarin/runtime.h]: It means someone outside of XI/XM may use this method (in this case the embeddinator), so it shouldn't be removed.

File: tools/linker/MobileResolveMainAssemblyStep.cs (hunk 2 above)
Comment (rolfbjarne) [inline tools/linker/MobileResolveMainAssemblyStep.cs]: Yeah, we only need to mark a single thing. I couldn't decide how to select that single thing in any way that made sense, so I just went for something I thought made sense: mark every public type.

File: tools/linker/MobileResolveMainAssemblyStep.cs (hunk 1 above)
Comment (rolfbjarne) [inline tools/linker/MobileResolveMainAssemblyStep.cs]: Fixed.

File: runtime/xamarin/runtime.h (hunk 3 above)
Comment (dalexsoto) [inline runtime/xamarin/runtime.h]: Can we make this more explicit in the source comment? Had the same question as Sebastien when I saw it. Just to help our future selves and anyone else 🙂

**Pull request merged**